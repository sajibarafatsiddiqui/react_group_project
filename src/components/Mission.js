import PropTypes from 'prop-types';
import Button from 'react-bootstrap/Button';
import { useDispatch } from 'react-redux';
import { toggleReservation } from '../redux/missions/missionsSlice';

const Mission = ({ missionProp }) => {
  const dispatch = useDispatch();
  let variant = 'outline-secondary';
  let text = 'Join Mission';
  let variantStatus = 'secondary';
  let statusText = 'NOT A MEMBER';
  if (missionProp.reserved) {
    variant = 'outline-danger';
    text = 'Leave Mission';
    variantStatus = 'success';
    statusText = 'Active Member';
  }

  return (
    <tr>
      <td><strong>{missionProp.missionName}</strong></td>
      <td>{missionProp.description}</td>
      <td className="align-middle"><Button variant={variantStatus} size="sm">{statusText}</Button></td>
      <td className="align-middle">
        <Button
          variant={variant}
          onClick={() => { dispatch(toggleReservation(missionProp.missionId)); }}
        >
          { text }
        </Button>
      </td>
    </tr>
  );
};
Mission.propTypes = {
  missionProp: PropTypes.shape({
    missionName: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    reserved: PropTypes.bool.isRequired,
    missionId: PropTypes.string.isRequired,
  }).isRequired,
};

export default Mission;
