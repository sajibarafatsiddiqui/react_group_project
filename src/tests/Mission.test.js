import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Mission from 'components/Mission';
import store from 'redux/store';
import { Table } from 'react-bootstrap';

describe('Tests for <Mission /> component', () => {
  const newMission = {
    missionId: '2',
    missionName: 'Mission 2',
    description: 'Mission description',
  };
  const testMission = (reserved) => (
    <Mission key={newMission.missionId} missionProp={{ ...newMission, reserved }} />

  );
  const wrapper = (component) => (
    <Provider store={store}>
      <Table>
        <tbody>

          {component}
        </tbody>
      </Table>
    </Provider>
  );
  it('Should render the content correctly', () => {
    render(wrapper(testMission(false)));
    expect(screen.getByText(newMission.missionName)).not.toBeNull();
    expect(screen.getByText(newMission.description)).not.toBeNull();
  });

  it('Should display the reservqe button when reserved is false', () => {
    render(wrapper(testMission(false)));
    expect(screen.queryByText('NOT A MEMBER')).not.toBeNull();
    expect(() => screen.getAllByText('Leave Mission')).toThrow();
  });

  it('Should display the cancel reservation button when reserved is true', () => {
    render(wrapper(testMission(true)));
    expect(screen.queryByText('Active Member')).not.toBeNull();
    expect(() => screen.getAllByText('Join Mission')).toThrow();
  });

  it('Should match the snapshot', () => {
    expect(renderer.create(wrapper(testMission(false))).toJSON()).toMatchSnapshot();
  });
});
