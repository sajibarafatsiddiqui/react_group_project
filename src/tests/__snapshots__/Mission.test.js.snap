// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <Mission /> component Should match the snapshot 1`] = `
<table
  className="table"
>
  <tbody>
    <tr>
      <td>
        <strong>
          Mission 2
        </strong>
      </td>
      <td>
        Mission description
      </td>
      <td
        className="align-middle"
      >
        <button
          className="btn btn-secondary btn-sm"
          disabled={false}
          type="button"
        >
          NOT A MEMBER
        </button>
      </td>
      <td
        className="align-middle"
      >
        <button
          className="btn btn-outline-secondary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Join Mission
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
